### PR TITLE
frontend: Add attach to pod option

### PR DESCRIPTION
**Pull request by ashu8912**
Acceptance Criteria:

- [x] Go to pods detail view and there is a attach header action
- [x] Click on the attach header action and you are able to connect to the container stdin, stdout, stderr whichever is available
- [x] Change the container and you get attached to the new container i/o/err stream now

### CODE REVIEW

File: frontend/src/components/common/Terminal.tsx [hunk 1]
```diff
@@ -76,7 +77,7 @@ export default function Terminal(props: TerminalProps) {
   const classes = useStyle();
   const [terminalContainerRef, setTerminalContainerRef] = React.useState<HTMLElement | null>(null);
   const [container, setContainer] = React.useState<string | null>(null);
-  const execRef = React.useRef<execReturn | null>(null);
+  const execAndAttachRef = React.useRef<execReturn | null>(null);
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:80]: Should be called `execOrAttachRef` I guess.

File: frontend/src/components/common/Terminal.tsx [hunk 2]
```diff
@@ -177,7 +178,6 @@ export default function Terminal(props: TerminalProps) {
       // On server error, don't set it as connected
       if (channel !== Channel.ServerError) {
         xtermc.connected = true;
-        console.debug('Terminal is now connected');
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:180]: Why remove this?

File: frontend/src/components/common/Terminal.tsx [hunk 3]
```diff
@@ -162,9 +163,9 @@ export default function Terminal(props: TerminalProps) {
     // The first byte is discarded because it just identifies whether
     // this data is from stderr, stdout, or stdin.
     const data = bytes.slice(1);
-    const text = decoder.decode(data);
+    let text = decoder.decode(data);
 
-    if (bytes.byteLength < 2) {
+    if (bytes.byteLength < 1) {
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:168]: Is this needed in the end if the first byte is just telling which channel to use?

File: frontend/src/components/common/Terminal.tsx [hunk 4]
```diff
@@ -197,8 +197,30 @@ export default function Terminal(props: TerminalProps) {
       shellConnectFailed(xtermc);
       return;
     }
-
-    xterm.write(text);
+    // replace \n with \r only if there is \r before \n
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:200]: I don't this this is what's being done on line 202.

File: frontend/src/components/common/Terminal.tsx [hunk 5]
```diff
@@ -197,8 +197,30 @@ export default function Terminal(props: TerminalProps) {
...
+    // give me the regular expression for the above
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:201]: Remove this comment.

File: frontend/src/components/common/Terminal.tsx [hunk 6]
```diff
@@ -197,8 +197,30 @@ export default function Terminal(props: TerminalProps) {
...
+    text.replace(/\r\n/g, '\r').replace(/\n/g, '\r');
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:202]: This is not doing anything.

File: frontend/src/components/common/Terminal.tsx [hunk 7]
```diff
@@ -197,8 +197,30 @@ export default function Terminal(props: TerminalProps) {
...
+    if (props.isAttach) {
+      // in case of attach if we didn't recieve any data from the process we should try to reconnect
+      if (!text) {
+        text = "If you don't see a command prompt, try pressing enter.";
+      }
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:207]: The terminal is not responding to any input, so we should remove this message as it's confusing.

File: frontend/src/components/common/Terminal.tsx [hunk 8]
```diff
@@ -197,8 +197,30 @@ export default function Terminal(props: TerminalProps) {
...
+      // in case of attach we should replace \n with \r\n because it's not formatted the way it is for exec so we have to manually do this
+      xterm.write(
+        text.replace(/(\r)?\n/g, function (match, p1) {
+          // If there is a \r before \n, replace \n with \r\n
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:211]: Is this really what's being done in line 213?

File: frontend/src/components/common/Terminal.tsx [hunk 9]
```diff
@@ -197,8 +197,30 @@ export default function Terminal(props: TerminalProps) {
...
+          if (p1 === '\r') {
+            return '';
+          }
+          // Otherwise, just return \r\n
+          else {
+            return '\r\n';
+          }
+        })
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:219]: I think this whole regex can be simplified by doing:
`text = text.replace(/\r\n/g, '\n').replace(/\n/g, '\r\n');`

Can you check if this is the case?

File: frontend/src/components/common/Terminal.tsx [hunk 10]
```diff
@@ -197,8 +197,30 @@ export default function Terminal(props: TerminalProps) {
...
+      );
+    } else {
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:221]: If we just reassign the text above (without writing it) and remove the else, the code gets simpler to read.

File: frontend/src/components/common/Terminal.tsx [hunk 11]
```diff
@@ -278,17 +300,28 @@ export default function Terminal(props: TerminalProps) {
...
       (async function () {
-        execRef.current = await item.exec(
-          container,
-          (items: ArrayBuffer) => onData(xtermRef.current!, items),
-          { command: [command], failCb: () => shellConnectFailed(xtermRef.current!) }
-        );
-
+        if (props.isAttach) {
+          xtermRef?.current?.xterm.writeln(
+            t(`Trying to attach to the container ${container}...`, { container }) + '\n'
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:306]: Please check how the translation functions are used with variables.

File: frontend/src/components/common/Terminal.tsx (hunk 7 above)
Comment (joaquimrocha) [reply frontend/src/components/common/Terminal.tsx:207]: Can we check if, once the attach is successful, we just print a message saying "Any new output for this container's process should be shown below."

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Good job. I tried it and it works but needs some improvements in terms of UX and the code.
Also, maybe the terminal should have a title "Attach: POD_NAME" instead of "Terminal: POD_NAME"

File: frontend/src/components/common/Terminal.tsx [hunk 12]
```diff
@@ -278,17 +287,28 @@ export default function Terminal(props: TerminalProps) {
...
       (async function () {
-        execRef.current = await item.exec(
-          container,
-          (items: ArrayBuffer) => onData(xtermRef.current!, items),
-          { command: [command], failCb: () => shellConnectFailed(xtermRef.current!) }
-        );
-
+        if (isAttach) {
+          xtermRef?.current?.xterm.writeln(
+            t(`Trying to attach to the container {{ container }}...`, { container }) + '\n'
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:293]: Shouldn't be a string literal as it'll result in warnings when running make i18n.

File: frontend/src/components/common/Terminal.tsx [hunk 13]
```diff
@@ -384,7 +404,11 @@ export default function Terminal(props: TerminalProps) {
       }}
       keepMounted
       withFullScreen
-      title={t('Terminal: {{ itemName }}', { itemName: item.metadata.name })}
+      title={
+        isAttach
+          ? t('Attach : {{ itemName }}', { itemName: item.metadata.name })
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:409]: Remove space between `Attach` and `:`, so it's consistent with the other title.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Just a couple minor changes left.

File: frontend/src/components/common/Terminal.tsx (hunk 12 above)
Comment (ashu8912) [reply frontend/src/components/common/Terminal.tsx]: updated

File: frontend/src/components/common/Terminal.tsx (hunk 13 above)
Comment (ashu8912) [reply frontend/src/components/common/Terminal.tsx]: updated

File: frontend/src/components/common/Terminal.tsx [hunk 14]
```diff
@@ -278,17 +288,28 @@ export default function Terminal(props: TerminalProps) {
...
       (async function () {
-        execRef.current = await item.exec(
-          container,
-          (items: ArrayBuffer) => onData(xtermRef.current!, items),
-          { command: [command], failCb: () => shellConnectFailed(xtermRef.current!) }
-        );
-
+        if (isAttach) {
+          xtermRef?.current?.xterm.writeln(
+            t('Trying to attach to the container {{ container }}...', { container }) + '\n'
```
Comment (joaquimrocha) [inline frontend/src/components/common/Terminal.tsx:294]: Please use … instead of ..., for consistency and better a11y.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Awesome!. Left one very small comment.

File: frontend/src/components/common/Terminal.tsx (hunk 14 above)
Comment (ashu8912) [reply frontend/src/components/common/Terminal.tsx]: udpated

Comment (joaquimrocha) [reply frontend/src/components/common/Terminal.tsx]: There was one missed `...` but I have fixed it.

**Review by joaquimrocha — APPROVED** [review]

**Pull request merged**